Make filters sticky in Dashboard page

gusaiani opened the pull request:
Fixes #11187

## How to Test

1. Visit dashboards that are tall enough to make the window scrollable.
2. Scroll
3. Filter should now be sticky at the top on scroll.

Try opening sidebars: filter, sharing, etc.

Try adding enough filters that the widget has more than one row of filters.

🎗️This PR is exclusively for regular view mode.
Will follow up with similar PRs serving edit, fullscreen modes if prioritized.

### Demo

![Metabase Scroll 1](https://user-images.githubusercontent.com/380816/128077075-7a4106a4-910f-499a-b991-326d72658748.gif)

gusaiani committed:
Create unit test file for Dashboard

gusaiani committed:
Start DashboardEmptyState

gusaiani committed:
Extract styled-component for Container

gusaiani committed:
Convert full width wrappers

gusaiani committed:
Extract DashboardStyled

gusaiani committed:
Trigger deploy

gusaiani committed:
Trigger deploy

gusaiani committed:
Add naive visual regression test

gusaiani committed:
Rename visual test file and descriptions

gusaiani committed:
Make visual test use 12 filters

gusaiani committed:
Trigger deploy

gusaiani commented:
@metabase-bot run visual tests

ariya commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:
```diff
@@ -75,9 +77,22 @@ export default class Dashboard extends Component {
   componentDidMount() {
     this.loadDashboard(this.props.dashboardId);
+
+    window.addEventListener("scroll", this.updateParametersWidgetStickiness, {
```
Is there any value of using `_.throttle` here?
I seem to recall that this was necessary for some pop-up resizing. /cc @daltojohnso

gusaiani replied:
@ariya @daltojohnso 

I will experiment with throttling, thanks for the suggestion.

gusaiani replied:
Done. 

Feels like a trade off:

for all browsers but FF it makes things very slightly worse from the best baseline.

for FF it makes things a heck of a lot better from a very low baseline.

Keeping the throttle at 24fps for now.

ariya replied:
Great job at identifying the potential issue with Firefox scrolling!

flamber reviewed (commented):
Much better. 
I'm still seeing the scroll-to-top-on-dropdown. And it seems like the stickiness is a little sluggish (this usually happens, when javascript has to do magic compared to a cleaner CSS approach, but React :man_shrugging:)

https://user-images.githubusercontent.com/1447303/129207608-9e81e5bd-585f-4fb1-bcff-c5918fd1072b.mp4

gusaiani commented:
@flamber Thank you very much.

I'm still full of basic blind spots about merely using Metabase.

Will see what I can do about the drop downs.

gusaiani committed:
Throttle scrolling and resizing

gusaiani commented:
@flamber From the latest commit, the dropdowns are still repositioned super slowly, it's quite jittery.

**But…**they don't go below the parameter bar anymore.

I've started to investigate how to render the dropdowns differently when the parameters are sticky.

Let's see how complicated it will be (pretty clearly it goes into Popover land).

flamber commented:
Yeah, I really wish we could replace our popover with something better - like react-select.

gusaiani committed:
Only update state if it should be different

gusaiani committed:
Extract function about paramsAndCards container

gusaiani committed:
Extract function to check if should be sticky

gusaiani committed:
Extract sticky functions to file

gusaiani committed:
Clean up ParameterWidget

mazameli commented:
Branch is looking good so far. One detail I'll point out that perhaps you just haven't gotten to yet is this border below the filters which should appear at the moment at which the filters become sticky, and disappear when they're no longer sticky:

![image](https://user-images.githubusercontent.com/2223916/129423148-69db58c5-292b-4228-a11a-ee5a46bfed71.png)

gusaiani commented:
> border below the filters

@mazameli Thanks for pointing it out, will update.

nemanjaglumac commented on frontend/test/metabase-visual/dashboard/parameters-widget.cy.spec.js:
```diff
@@ -0,0 +1,62 @@
+  type: "category",
+};
+
+const filters = new Array(12).fill(filter);
```
I like this idea, but it might need some tweaking, depending on your end goal.
We're filling this array 12x with absolutely the same filter (because of the hard coded `id` key). It then creates 12 identical filters, which is not even possible to achieve IRL.

Maybe you intentionally wanted to "hack" it this way simply to populate Ui with 12 filters (because we don't actually care about their functionality - we just want to make sure they remain sticky)?

gusaiani replied:
@nemanjaglumac 

Yes, as you point out, the idea was just to populate with any 12 filters and test the behaviour of their container.

Happy to change this and follow your instructions if you think we should be specific about these filters.

nemanjaglumac replied:
> Happy to change this and follow your instructions if you think we should be specific about these filters.

Nah, this is more than good for this use case. Just wanted to check the intent behind it.
Btw, in that case you can even remove the part where you connect the filter to the card. Still need to make that card tall enough, though. This is optional - don't waste time if you don't want to.

gusaiani committed:
Unit test stickyParameters

gusaiani commented:
@flamber, I have just a detail to add (border-bottom when sticky) and try to break it by using it.

If you would like to try it out again and report problems, I will be grateful.

flamber reviewed (commented):
This feels great :100: - not noticing any problems anywhere.

gusaiani committed:
Add borderBottom to filters when sticky

gusaiani committed:
Detail comment a bit more.

gusaiani commented:
@metabase-bot run visual tests

nemanjaglumac commented on frontend/test/metabase-visual/dashboard/parameters-widget.cy.spec.js:
```diff
@@ -0,0 +1,62 @@
+    restore();
+    cy.signInAsAdmin();
+
+    cy.createQuestionAndDashboard({ questionDetails }).then(
+      ({ body: { id, card_id, dashboard_id } }) => {
+        cy.request("PUT", `/api/dashboard/${dashboard_id}`, {
+          parameters: filters,
+        });
+
+        // Connect filter to the card
+        cy.request("PUT", `/api/dashboard/${dashboard_id}/cards`, {
+          cards: [
+            {
+              id,
+              card_id,
+              row: 0,
+              col: 0,
+              sizeX: 12,
+              sizeY: 32,
+              visualization_settings: {},
+              parameter_mappings: [
+                {
+                  parameter_id: "ad1c877e",
+                  card_id,
+                  target: ["dimension", ["field-id", PRODUCTS.CATEGORY]],
+                },
+              ],
+            },
+          ],
+        });
+
+        cy.visit(`/dashboard/${dashboard_id}`);
+      },
+    );
```
Heads-up:
Now that the https://github.com/metabase/metabase/pull/17493 is merged, you can simplify this part greatly.

We don't care about connecting that filter to the card, right? So we can omit `parameter_mappings` so the only thing left to do is to resize the card:

```javascript
cy.createQuestionAndDashboard({ questionDetails }).then(
  ({ body: oldCard }) => {
    const { dashboard_id } = oldCard;

    cy.request("PUT", `/api/dashboard/${dashboard_id}`, {
      parameters: filters,
    });

    const updatedSize = {
      sizeX: 12,
      sizeY: 32,
    };

    cy.editDashboardCard(oldCard, updatedSize);

    cy.visit(`/dashboard/${dashboard_id}`);
  },
);
```

It's absolutely not necessary to update this test. Just wanted to give you an example in case it looks interesting :)

gusaiani committed:
Update visual test

gusaiani committed:
Update e2e test

gusaiani committed:
Trigger deploy

nemanjaglumac committed:
Fix failing Cypress test

- The test was failing because the card was too small and Cypress couldn't click the "eye" icon
- `cy.createNativeQuestionAndDashboard` has some sane defaults for the dashboard card size which fixes this

nemanjaglumac commented:
I'm hitting the following error in the console:
```
stickyParameters.js:43 Uncaught TypeError: Cannot read property 'offsetHeight' of null
    at updateParametersAndCardsContainerStyle (stickyParameters.js:43)
    at updateParametersWidgetStickiness (stickyParameters.js:12)
    at Dashboard.jsx:92
    at throttled (throttle.js:32)
```

```javascript
// frontend/src/metabase/dashboard/components/Dashboard/stickyParameters.js:43
const { offsetHeight } = dashboard.parametersWidgetRef;
```

gusaiani commented:
@nemanjaglumac 🤔 thanks, will check again

nemanjaglumac reviewed (approved):
Apart from the console error I stumbled upon, this looks and works GREAT!
Nice work, @gusaiani 🎉

gusaiani commented:
@nemanjaglumac Thank you very much 🙂 Will clean up the console error on Monday first thing.

nemanjaglumac commented:
Confirming that this error doesn't exist on the latest `master`.
I could've caught some weird state when switching branches when I reported it.
